external_parser/multistep_example_joiner.cc

```diff
@@ -0,0 +1,246 @@
+                        event.context()->size());
+
+  if (_vw->audit || _vw->hash_inv) {
+    VW::template read_line_json<true>(
+        *_vw, examples, const_cast<char *>(line_vec.c_str()),
+        reinterpret_cast<VW::example_factory_t>(&VW::get_unused_example), _vw);
+  } else {
+    VW::template read_line_json<false>(
+        *_vw, examples, const_cast<char *>(line_vec.c_str()),
+        reinterpret_cast<VW::example_factory_t>(&VW::get_unused_example), _vw);
+  }
+  return joined_event(event_meta.timestamp, std::move(meta), std::move(data), line_vec, event.model_id() ? event.model_id()->c_str() : "N/A");
+}
+
+void try_set_label(const joined_event &je, float reward,
+                                   v_array<example *> &examples) {
+  if (je.interaction_data.actions.empty()) {
+    VW::io::logger::log_error("missing actions for event [{}]",
+                              je.interaction_data.eventId);
+    return;
+  }
+
+  if (je.interaction_data.probabilities.empty()) {
+    VW::io::logger::log_error("missing probabilities for event [{}]",
+                              je.interaction_data.eventId);
+    return;
+  }
+
+  if (std::any_of(je.interaction_data.probabilities.begin(),
+                  je.interaction_data.probabilities.end(),
+                  [](float p) { return std::isnan(p); })) {
+    VW::io::logger::log_error(
+        "distribution for event [{}] contains invalid probabilities",
+        je.interaction_data.eventId);
+  }
+
+  int index = je.interaction_data.actions[0];
+  auto action = je.interaction_data.actions[0];
+  auto cost = -1.f * reward;
+  auto probability = je.interaction_data.probabilities[0];
+  auto weight = 1.f - je.interaction_data.probabilityOfDrop;
+
+  examples[index]->l.cb.costs.push_back({cost, action, probability});
+  examples[index]->l.cb.weight = weight;
+}
+
+bool multistep_example_joiner::process_joined(v_array<example *> &examples) {
+  if (!_sorted) {
+    populate_order();
+  }
+  const auto& id = _order.front();
+
+  const auto& interactions = _interactions[id];
+  if (interactions.size() != 1) {
+    return -1;
+  }
+  const auto& interaction = interactions[0];
+  auto joined = process_interaction(interaction, examples);
+
+  const auto outcomes = _outcomes[id];
+  for (const auto& o: outcomes) {
+    joined.outcome_events.push_back(process_outcome(o));
+  }
+  for (const auto& o: _episodic_outcomes) {
+    joined.outcome_events.push_back(process_outcome(o));
+  }
+  const auto reward = _reward_calculation(joined);
+  try_set_label(joined, reward, examples);
+  
+  // add an empty example to signal end-of-multiline
+  examples.push_back(&VW::get_unused_example(_vw));
+  _vw->example_parser->lbl_parser.default_label(&examples.back()->l);
+  examples.back()->is_newline = true;
+
+  _order.pop();
+  return true;
+}
+
+bool multistep_example_joiner::processing_batch() {
+  return _sorted && !_order.empty();
+}
+
+void multistep_example_joiner::on_new_batch() {
```

Comment:
could `on_new_batch` and maybe `on_batch_read` be operations that are done at the end of every batch instead? that way they don't need to be introduced at the binary parser level and each batch can just cleanup

Reply:
on second thought these could be nice for future potential implementations as a clear indication that we are switching to a new batch